KIPs/kip-162.md

```diff
@@ -0,0 +1,204 @@
+
+```py
+def EffectiveGasPrice(header: Header, tx: Transaction):
+    if header.number >= DRAGON_FORK_BLOCK_NUMBER:
+      # Transaction type 2 has {maxFeePerGas, maxPriorityFeePerGas} fields
+      # Other transactions only have {gasPrice} field
+      if tx.type == 2:
+          maxFeePerGas = tx.maxFeePerGas
+          maxPriorityFeePerGas = tx.maxPriorityFeePerGas
+      else:
+          # makes EffectiveGasPrice() = tx.gasPrice
+          maxFeePerGas = tx.gasPrice
+          maxPriorityFeePerGas = tx.gasPrice
+      return min(header.baseFeePerGas + maxPriorityFeePerGas, maxFeePerGas)
+
+    elif header.number >= MAGMA_FORK_BLOCK_NUMBER:
+      return header.baseFeePerGas
+
+    else:
+      # Note: before Magma hardfork, transactions must satisfy
+      # tx.gasPrice == config.UnitPrice or tx.maxFeePerGas == config.UnitPrice.
+      # Therefore EffectiveGasPrice is essentially config.UnitPrice.
+      if tx.type == 2:
+          return tx.maxFeePerGas
+      else:
+          return tx.gasPrice
+
+def EffectivePriorityFeePerGas(header: Header, tx: Transaction):
+    if header.number >= DRAGON_FORK_BLOCK_NUMBER:
+        effectiveGasPrice = EffectiveGasPrice(header, tx)
+        return effectiveGasPrice - header.baseFeePerGas
+    else:
+        return 0
+```
+
+### KIP-82 reward scheme
+
+The [KIP-82](https://github.com/klaytn/kips/blob/main/KIPs/kip-82.md) reward scheme is slightly modified. The `get_total_fee` now accounts for effective gas price. Otherwise remains the same.
+
+```py
+def get_total_fee(header: Header, txs: list[Transaction], receipts: list[Receipt]):
+    # Note: EffectiveGasPrice handles hardfork
+    totalFee = 0
+    for i in range(len(txs)):
+        totalFee += EffectiveGasPrice(header, txs[i]) * receipts[i].gasUsed
+    return totalFee
+```
+
+### EVM GASPRICE opcode
+
+The `GASPRICE` (0x3a) opcode returns the effective gas price of the currently executing transaction.
+
+## Transaction pool
+
+### Transaction ordering
+
+Block transaction ordering depends on client implementation as the ordering is not checked by block validators. However, under the KIP-82 reward scheme, a proposer should prioritize high priority fee transactions for the maximum proposer reward.
+
+Block proposers are recommended to order transactions with descending effective gas price, and the same effective gas price are sorted by transaction arrival time. This discourages the transaction spamming where a sender sends many transactions hoping that at least one is included in the block. Instead, the sender would pay a higher priority fee to ensure the transaction be included in the block.
+
+### Txpool management
+
+A transaction pool should not accept underpriced transactions. A transaction is underpriced if its `tx.gasPrice` or `tx.maxFeePerGas` field is lower than the baseFeePerGas. The transaction pool can use last block's baseFeePerGas or the baseFeePerGas calculated for next block.
```

Comment:
> use last block's baseFeePerGas or the baseFeePerGas calculated for next block

Which is it in Klaytn? and could you add this for clarification?

Reply:
Fixed in 344af4e. It's the calculated baseFeePerGas of the next (pending) block. The `pool.gasPrice`, which holds the required minimum price, is updated (https://github.com/klaytn/klaytn/blob/v1.12.1/blockchain/tx_pool.go#L488) whenever the txpool receives the ChainHeadEvent (https://github.com/klaytn/klaytn/blob/v1.12.1/blockchain/tx_pool.go#L307).